### ISSUE TITLE
Support for multiple stages in a single CI run

### ISSUE THREAD

**Issue by testphreak** [anshooarora/extentx#3]
Hi @anshooarora , I didn't find Gitter very sticky, especially for discussions around enhancements. So I am creating an enhancement request here to make it easier to track for you and me.

Hopefully with ExtentX still in early development, the underlying data model can be changed to link multiple stages in a single run and merge those results.

Here's a copy of our discussion on ExtentReports issues page and Gitter for reference:

https://github.com/anshooarora/extentreports/issues/345

testphreak Mar 04 19:28
Hi Anshoo! Did you get a chance to think through a potential solution for the scenario I outlined for ExtentX?
It would be huge!

anshooarora Mar 04 19:52
I did, but unable to think of an elegant way to solve for this
since there will be a new Extent instance, it will create new associations for Report -> Tests -> Nodes/Logs/Categories etc.
The only thing I can think of is passing the ID of the 1st run for each retry

testphreak Mar 04 23:15
Yes, that's what I was thinking. ExtentX keeps incrementing the overall run count by 1 with every run, but allows the user to set numberOfStages or something similar to let ExtentX know when it needs to stop counting subsequent runs as stages and start a brand new run. ExtentX would have to pass the ID of the 1st stage/run with each retry. Once numberOfStages reaches the user defined limit, ExtentX would merge the results of those stages. I hope that sounds feasible.

**Comment by anshooarora** [comment]
Do you use `IRetryAnalyzer` to re-run failed tests?  I am looking at [this](http://www.seleniumeasy.com/testng-tutorials/execute-only-failed-test-cases-using-iretryanalyzer) - is this the approach I should follow?

**Comment by testphreak** [comment]
No I do not use `IRetryAnalyzer`. I use out of the box solution - testng-failed.xml file that gets generated in the output directory when tests fail in a suite.

More info under "Rerunning failed tests" - http://testng.org/doc/documentation-main.html

**Comment by anshooarora** [comment]
Does the 2nd stage in your scenario run after the 1st is complete and all the reports created, correct?  Same for the 3rd, will run after the 2nd is completely executed?

**Comment by testphreak** [comment]
Yes, that's correct. Each stage runs sequentially using the previous stage's testng-failed.xml and creates it's own report.

**Comment by anshooarora** [comment]
Okay, so you are then using `replaceExisting = true` and appending tests to the same report?

**Comment by testphreak** [comment]
No, each stage saves it's own report. Here's a screenshot of a Bamboo run, to explain better. Ignore the word ReportNG in the name of the artifact (essentially ExtentReports Test Results).

![image](https://cloud.githubusercontent.com/assets/10567800/14158251/fe178b48-f683-11e5-955f-2daff7c9ca2b.png)

**Comment by anshooarora** [comment]
The only elegant way I can think of is if you can pass the id of the 1st try or the its filepath.  With ID, you can query from the db and associate with your run or let extent parse the file for it.

**Comment by testphreak** [comment]
Hi Anshoo, which id are you referring to - the id of the Bamboo stage or an ExtentReports id? Would be great if you could elaborate further as the second line wasn't clear to me either.

**Comment by anshooarora** [comment]
This is the mongo database id.  Since all data is stores in a mongodb instance, if you can provide the report location or its ID, all info will be applied to the same report.

**Comment by testphreak** [comment]
Will there be an API exposed which I can call to get the database Id? Still trying to wrap my head around how it would all work together i.e. CI (Bamboo/Jenkins etc) with ExtentX (NodeJS and MongoDB). Is there an ExtentX sequence diagram you can come up with when you have some time? Just that I'm not sure how ExtentX would work with retries either with stages (testng-failed.xml) or a single stage with `IRetryAnalyzer` and come up with a single report.

So, in my example there were 6 stages in the same run. After the first stage completes, I call an API or write a [bash script](http://stackoverflow.com/a/9558954/1093019) to get the database Id. Do I then pass this id to the second stage of the same run and so on?

This is great work Anshoo, just want to help make it truly useful for a lot of automation projects!

**Comment by anshooarora** [comment]
It will be replacing ExtentMerge - so all your reports created will be stored on a central server to view.  There are other features such as viewing by project, search and analysis.

If you are already using Bamboo which does a pretty good job at this already, I don't see ExtentX providing much value as Bamboo takes care of some of these aspects already.

**Comment by testphreak** [comment]
No Bamboo will not do what this does, since you cannot do meaningful analysis of your runs over a certain time with a single dashboard. Also, Bamboo output is generally deleted after a month or so (at least where I work).

So yes, I will need to use a test tracking and analysis tool like ExtentX to gather results of all my runs. Could you suggest what I would need to do get the Mongo database id? Also, will there be a way in ExtentX to not report results for every stage of the run, but collect the results from all the stages and form 1 report for the entire run.

If the multiple stages scenario implementation is turning out to be complicated, are you looking into supporting IRetryAnalyzer for ExtentX? I can look into modifying my framework to use IRetryAnalyzer instead of running in multiple stages in Bamboo.

The reason why I liked the multiple stage approach using testng-failed.xml is I know after each try how many failed without having to wait for the entire run to finish as I would likely need to do with IRetryAnalyzer which will essentially be a single run with multiple tries in the same run.
